### PR TITLE
feat: update MerklePayoutStrategyImplementation

### PR DESCRIPTION
**Pull request by bhargavaparoksham**
##### Description

Implement MerklePayoutStrategyImplementation

Refer https://github.com/gitcoinco/merkle_payouts/blob/main/contracts/MerklePayout.sol as reference

##### Refers/Fixes

refs #1110

### CODE REVIEW

File: packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol [hunk 1]
```diff
@@ -71,7 +83,48 @@ contract MerklePayoutStrategyImplementation is IPayoutStrategy, Initializable {
    *
    * @param encodedDistribution encoded distribution
    */
-  function payout(bytes[] memory encodedDistribution) external override isRoundContract payable {
-    // TODO: Add payout logic
+  function batchPayout(Distribution[] calldata _distributions) external override isRoundContract payable {
```
Comment (thelostone-mc) [inline packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol:86]: Should call come via Round / directly to payout

File: packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol [hunk 2]
```diff
@@ -71,7 +83,48 @@ contract MerklePayoutStrategyImplementation is IPayoutStrategy, Initializable {
...
+    uint256 _amount = _distribution.amount;
+    bytes32[] calldata _merkleProof = _distribution.merkleProof;
+
+    require(!hasDistributed(_index), "MerklePayout: Funds already distributed.");
+
+    bytes32 node = keccak256(abi.encodePacked(_index, _grantee, _amount));
+    require(MerkleProof.verify(_merkleProof, merkleRoot, node), "MerklePayout: Invalid proof.");
+
+    _setDistributed(_index);
+    token.safeTransfer(_claimee, _amount);
```
Comment (thelostone-mc) [inline packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol:108]: handle native token transfer

File: packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol [hunk 3]
```diff
@@ -71,7 +83,48 @@ contract MerklePayoutStrategyImplementation is IPayoutStrategy, Initializable {
...
    */
-  function payout(bytes[] memory encodedDistribution) external override isRoundContract payable {
-    // TODO: Add payout logic
+  function batchPayout(Distribution[] calldata _distributions) external override isRoundContract payable { 
+
+    for (uint256 i = 0; i < _distributions.length; ++i) {
+        distribute(_distributions[i]);
+    }
+
+    emit BatchPayoutTriggered(msg.sender);
```
Comment (thelostone-mc) [inline packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol:92]: discuss if this needed ?

File: packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol [hunk 4]
```diff
@@ -71,7 +91,48 @@ contract MerklePayoutStrategyImplementation is IPayoutStrategy, Initializable {
...
+  function batchPayout(Distribution[] calldata _distributions) external override isRoundContract payable { 
+
+    for (uint256 i = 0; i < _distributions.length; ++i) {
+        _distribute(_distributions[i]);
+    }
+
+    emit BatchPayoutTriggered(msg.sender);
+  }
+
+  function _distribute(Distribution calldata _distribution) internal {
```
Comment (thelostone-mc) [inline packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol:103]: should this be public ?

Comment (hmrtn) [reply packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol]: no - unless we want to support individual payouts

File: packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol (hunk 2 above)
Comment (hmrtn) [reply packages/contracts/contracts/payoutStrategy/MerklePayoutStrategy/MerklePayoutStrategyImplementation.sol]: We first must have clear where the balances live!

**Pull request merged**